File: backends/ruby-rails/app/usecase/client/interactor.rb
# frozen_string_literal: true
#
# クライアントユースケースの実装モジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

require "openssl"
require "securerandom"
require "base64"

module UseCase
  module Client
    # クライアントに関するユースケースの実装クラスです。
    # @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
    class Interactor
      # @param repo [Domain::Client::Repository] クライアントリポジトリ
      def initialize(repo)
        @repo = repo
      end

      def find_by_condition(dto)
        cond = Domain::Client::Condition.new(
          keyword:    dto.keyword,
          start_from: dto.start_from,
          start_to:   dto.start_to,
          statuses:   dto.statuses,
        )
        @repo.find_by_condition(cond)
      end

      def find_by_id(id)
        entity = @repo.find_by_id(id)
        entity_to_detail_vo(entity)
      end

      def store(dto)
        key         = OpenSSL::PKey::RSA.new(4096)
        priv_pem    = key.to_pem
        pub_pem     = key.public_key.to_pem
        fingerprint = "SHA256:#{Base64.strict_encode64(OpenSSL::Digest::SHA256.digest(key.public_key.to_der)).delete('=')}"
        access_token = SecureRandom.hex(32)
        identifier   = SecureRandom.hex(16)
        now          = Time.current

        entity = Domain::Client::Entity.new(
          id:           nil,
          name:         dto.name,
          identifier:   identifier,
          post_code:    dto.post_code,
          pref:         dto.pref,
          city:         dto.city,
          address:      dto.address,
          building:     dto.building,
          tel:          dto.tel,
          email:        dto.email,
          access_token: access_token,
          private_key:  priv_pem,
          public_key:   pub_pem,
          fingerprint:  fingerprint,
          status:       Domain::Client::Status::INACTIVE,
          start_at:     nil,
          stop_at:      nil,
          created_at:   now,
          created_by:   dto.executor_id,
          updated_at:   now,
          updated_by:   dto.executor_id,
          deleted_at:   nil,
          deleted_by:   nil,
          version:      1,
        )
        saved = @repo.save(entity)
        Domain::Client::StoreResultVo.new(
          id:           saved.id,
          name:         saved.name,
          email:        saved.email,
          access_token: saved.access_token,
        )
      end

      def update(dto)
        entity = @repo.find_by_id(dto.id)
        now    = Time.current

        updated         = entity.dup
        updated.name    = dto.name    if dto.name
        updated.post_code = dto.post_code if dto.post_code
        updated.pref    = dto.pref    if dto.pref
        updated.city    = dto.city    if dto.city
        updated.address = dto.address if dto.address
        updated.building = dto.building if dto.building
        updated.tel     = dto.tel     if dto.tel
        updated.email   = dto.email   if dto.email

        if dto.status && dto.status != entity.status
          updated.status  = dto.status
          updated.start_at = now if dto.status == Domain::Client::Status::ACTIVE && entity.start_at.nil?
        end

        updated.updated_at = now
        updated.updated_by = dto.executor_id

        saved = @repo.save(updated)
        entity_to_detail_vo(saved)
      end

      def destroy(dto)
        @repo.soft_delete(dto.id, dto.executor_id, dto.version)
        nil
      end

      def find_by_access_token(token)
        @repo.find_by_access_token(token)
      end

      def find_by_identifier(identifier)
        @repo.find_by_identifier(identifier)
      end

      private

      def entity_to_detail_vo(e)
        Domain::Client::DetailVo.new(
          id:         e.id,
          name:       e.name,
          identifier: e.identifier,
          post_code:  e.post_code,
          pref:       e.pref,
          city:       e.city,
          address:    e.address,
          building:   e.building,
          tel:        e.tel,
          email:      e.email,
          status:     e.status,
          start_at:   e.start_at,
          stop_at:    e.stop_at,
          created_at: e.created_at,
          updated_at: e.updated_at,
        )
      end
    end
  end
end

File: backends/go-echo/internal/usecase/client/interactor.go
// Package client はクライアントユースケースを提供します。
package client

import (
	domclient "authorization-go-echo/internal/domain/client"
	"authorization-go-echo/pkg/apperror"
	"bytes"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"encoding/pem"
	"fmt"
	"math/big"
	"strings"
	"time"
)

// Interactor はクライアントユースケースの実装です。
type Interactor struct {
	repo domclient.Repository
}

// NewInteractor は Interactor を生成します。
func NewInteractor(repo domclient.Repository) *Interactor {
	return &Interactor{repo: repo}
}

// AuthenticateByToken はアクセストークンでクライアントを認証します。
func (uc *Interactor) AuthenticateByToken(token string) (bool, error) {
	c, err := uc.repo.FindByAccessToken(&domclient.Client{AccessToken: token})
	if err != nil {
		return false, err
	}
	return c != nil, nil
}

// FindByCondition は条件に合うクライアント一覧を取得します。
func (uc *Interactor) FindByCondition(cond domclient.Condition) ([]*domclient.ListItem, error) {
	clients, err := uc.repo.FindByCondition(cond)
	if err != nil {
		return nil, err
	}
	items := make([]*domclient.ListItem, 0, len(clients))
	for _, c := range clients {
		items = append(items, clientToListItem(c))
	}
	return items, nil
}

// CountByCondition は検索条件に合致するクライアントの総件数を返します。
func (uc *Interactor) CountByCondition(cond domclient.Condition) (int, error) {
	return uc.repo.CountByCondition(cond)
}

// FindByID はIDでクライアントを取得します。
func (uc *Interactor) FindByID(dto FindByIDDto) (*domclient.DetailVo, error) {
	c, err := uc.repo.FindByID(&domclient.Client{ID: dto.ID})
	if err != nil {
		return nil, err
	}
	if c == nil {
		return nil, apperror.NotFound("client_not_found")
	}
	return clientToDetailVo(c), nil
}

// Store はクライアントを新規登録します。
func (uc *Interactor) Store(dto StoreDto) (*domclient.StoreVo, error) {
	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
	if err != nil {
		return nil, fmt.Errorf("rsa key generation: %w", err)
	}

	privPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
	})

	pubDER, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
	if err != nil {
		return nil, err
	}
	pubPEM := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: pubDER})

	fingerprint := rsaFingerprint(&privateKey.PublicKey)

	tokenBytes := make([]byte, 32)
	if _, err = rand.Read(tokenBytes); err != nil {
		return nil, err
	}
	accessToken := hex.EncodeToString(tokenBytes)

	now := time.Now()
	c := &domclient.Client{
		Name:        dto.Name,
		Identifier:  generateIdentifier(),
		PostCode:    dto.PostCode,
		Pref:        dto.Pref,
		City:        dto.City,
		Address:     dto.Address,
		Building:    dto.Building,
		Tel:         dto.Tel,
		Email:       dto.Email,
		AccessToken: accessToken,
		PrivateKey:  string(privPEM),
		PublicKey:   string(pubPEM),
		Fingerprint: fingerprint,
		Status:      domclient.StatusInactive,
		CreatedAt:   now,
		UpdatedAt:   now,
		CreatedBy:   &dto.ExecutorID,
		UpdatedBy:   &dto.ExecutorID,
	}

	saved, err := uc.repo.Save(c)
	if err != nil {
		return nil, err
	}
	return &domclient.StoreVo{
		ID:          saved.ID,
		Name:        saved.Name,
		Identifier:  saved.Identifier,
		Email:       saved.Email,
		AccessToken: saved.AccessToken,
	}, nil
}

// Update はクライアント情報を更新します。
func (uc *Interactor) Update(dto UpdateDto) (*domclient.DetailVo, error) {
	c, err := uc.repo.FindByIDIncludeDeleted(&domclient.Client{ID: dto.ID})
	if err != nil || c == nil {
		return nil, apperror.NotFound("client_not_found")
	}
	if c.Version != dto.Version {
		return nil, apperror.Conflict("optimistic_lock")
	}

	if dto.Name != nil {
		c.Name = *dto.Name
	}
	if dto.PostCode != nil {
		c.PostCode = *dto.PostCode
	}
	if dto.Pref != nil {
		c.Pref = *dto.Pref
	}
	if dto.City != nil {
		c.City = *dto.City
	}
	if dto.Address != nil {
		c.Address = *dto.Address
	}
	if dto.Building != nil {
		c.Building = *dto.Building
	}
	if dto.Tel != nil {
		c.Tel = *dto.Tel
	}
	if dto.Email != nil {
		c.Email = *dto.Email
	}

	if dto.Status != nil {
		c.Status = *dto.Status
		now := time.Now()
		if *dto.Status == domclient.StatusActive && c.StartAt == nil {
			c.StartAt = &now
			c.StopAt = nil
		}
		if *dto.Status == domclient.StatusSuspended {
			c.StopAt = &now
		}
	}

	now := time.Now()
	c.UpdatedAt = now
	c.UpdatedBy = &dto.ExecutorID

	saved, err := uc.repo.Save(c)
	if err != nil {
		return nil, err
	}
	return clientToDetailVo(saved), nil
}

// GetQr はidentifierでクライアントを取得し、QRコード用データを返します。
func (uc *Interactor) GetQr(dto FindByIdentifierDto) (*domclient.QrVo, error) {
	c, err := uc.repo.FindByIdentifier(&domclient.Client{Identifier: dto.Identifier})
	if err != nil {
		return nil, err
	}
	if c == nil {
		return nil, apperror.NotFound("client_not_found")
	}
	return &domclient.QrVo{
		Identifier:  c.Identifier,
		DeeplinkURL: "authgateway://clients/" + c.Identifier + "/info",
	}, nil
}

// GetInfo はidentifierでクライアント情報を返します。
func (uc *Interactor) GetInfo(dto FindByIdentifierDto) (*domclient.InfoVo, error) {
	c, err := uc.repo.FindByIdentifier(&domclient.Client{Identifier: dto.Identifier})
	if err != nil {
		return nil, err
	}
	if c == nil {
		return nil, apperror.NotFound("client_not_found")
	}
	return &domclient.InfoVo{
		Identifier: c.Identifier,
		Name:       c.Name,
		Status:     c.Status,
	}, nil
}

// Start はidentifierでクライアントを取得し、利用開始処理を行ってアクセストークンを返します。
func (uc *Interactor) Start(dto FindByIdentifierDto) (*domclient.StartVo, error) {
	c, err := uc.repo.FindByIdentifier(&domclient.Client{Identifier: dto.Identifier})
	if err != nil {
		return nil, err
	}
	if c == nil {
		return nil, apperror.NotFound("client_not_found")
	}
	if c.Status != domclient.StatusActive {
		now := time.Now()
		c.Status = domclient.StatusActive
		if c.StartAt == nil {
			c.StartAt = &now
		}
		c.StopAt = nil
		c.UpdatedAt = now
		c, err = uc.repo.Save(c)
		if err != nil {
			return nil, err
		}
	}
	return &domclient.StartVo{AccessToken: c.AccessToken}, nil
}

// Stop はidentifierでクライアントを取得し、利用停止処理を行います。
func (uc *Interactor) Stop(dto FindByIdentifierDto) error {
	c, err := uc.repo.FindByIdentifier(&domclient.Client{Identifier: dto.Identifier})
	if err != nil {
		return err
	}
	if c == nil {
		return apperror.NotFound("client_not_found")
	}
	if c.Status == domclient.StatusActive {
		now := time.Now()
		c.Status = domclient.StatusSuspended
		c.StopAt = &now
		c.UpdatedAt = now
		if _, err = uc.repo.Save(c); err != nil {
			return err
		}
	}
	return nil
}

// Destroy はクライアントを論理削除します。
func (uc *Interactor) Destroy(dto DestroyDto) error {
	c, err := uc.repo.FindByIDIncludeDeleted(&domclient.Client{ID: dto.ID})
	if err != nil || c == nil {
		return apperror.NotFound("client_not_found")
	}
	if c.Version != dto.Version {
		return apperror.Conflict("optimistic_lock")
	}

	now := time.Now()
	c.Status = domclient.StatusClosed
	c.UpdatedAt = now
	c.UpdatedBy = &dto.ExecutorID
	if _, err = uc.repo.Save(c); err != nil {
		return err
	}

	return uc.repo.SoftDelete(&domclient.Client{ID: dto.ID, DeletedBy: &dto.ExecutorID})
}

func clientToListItem(c *domclient.Client) *domclient.ListItem {
	return &domclient.ListItem{
		ID:        c.ID,
		Name:      c.Name,
		Status:    c.Status,
		StartAt:   c.StartAt,
		StopAt:    c.StopAt,
		CreatedAt: c.CreatedAt,
		UpdatedAt: c.UpdatedAt,
	}
}

func clientToDetailVo(c *domclient.Client) *domclient.DetailVo {
	return &domclient.DetailVo{
		ID:         c.ID,
		Name:       c.Name,
		Identifier: c.Identifier,
		PostCode:   c.PostCode,
		Pref:       c.Pref,
		City:       c.City,
		Address:    c.Address,
		Building:   c.Building,
		Tel:        c.Tel,
		Email:      c.Email,
		Status:     c.Status,
		StartAt:    c.StartAt,
		StopAt:     c.StopAt,
		CreatedAt:  c.CreatedAt,
		UpdatedAt:  c.UpdatedAt,
	}
}

func rsaFingerprint(pub *rsa.PublicKey) string {
	buf := new(bytes.Buffer)

	writeBytes := func(b []byte) {
		_ = binary.Write(buf, binary.BigEndian, uint32(len(b)))
		buf.Write(b)
	}

	writeBytes([]byte("ssh-rsa"))
	writeBytes(big.NewInt(int64(pub.E)).Bytes())
	writeBytes(pub.N.Bytes())

	h := sha256.Sum256(buf.Bytes())
	return "SHA256:" + strings.TrimRight(base64.StdEncoding.EncodeToString(h[:]), "=")
}

func generateIdentifier() string {
	b := make([]byte, 8)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}
